designs/dynamic-bundle.md

```diff
@@ -0,0 +1,267 @@
+	kustomizecommonv1.Plugin{},
+	golangv3.Plugin{},
+)
+```
+
+In the above example a plugin `Bundle` is created that ensures that the `kustomize/v1` plugin is run before the `go/v3` plugin.
+ 
+With the addition of Phase 2 Plugins on the way, users using the `--plugins` flag to utilize external plugins will not get the benefits of using some default functionality that previous Kubebuilder users using the default `go/v3` plugin gained via the `kustomize/v1` plugin.
+
+## Proposal
+The proposed solution is to create a new form of `Bundle` that is more dynamic and can be used to create a default bundle that contains sensible default plugins that can be used with all plugins. This could be overriden by a command line flag.
+
+## User Stories
+- As a Kubebuilder maintainer I would like to have Kubebuilder implement sane default plugins that always run before and/or after plugins, specifying this once and it applying to all uses
+
+- As a Kubebuilder user I expect Kubebuilder to have default functionality to make my operator development easier
+
+- As a Kubebuilder user I would like to be able to override the default Kubebuilder plugin chain, scaffolding only exactly what I feel I need
+
+## Goals
+- `kubebuilder` is able to define sane defaults for plugins that should always be run before and/or after user specified plugins
+
+- A `Bundle` no longer has to be created for each plugin to implement sane defaults
+
+## Examples
+Context: Kubebuilder defines that the `kustomize/v1` plugin should always run before any other plugin. Kubebuilder also defines the default plugin to be run if a user does not specify plugins via the `--plugins` flag is `go/v3`
+
+- `kubebuilder init`
+	- Should scaffold using the `kustomize/v1` plugin and the `go/v3` plugin
+- `kubebuilder init --plugins other.plugin/v3`
+	- Should scaffold using the `kustomize/v1` plugin and the `other.plugin/v3` plugin
+- `kubebuilder init --plugins other.plugin/v3 --override-default-plugin-chain`
+	- Should scaffold using only the `other.plugin/v3` plugin
+
+## Implementation Details
+### Interface
+A new interface called `DynamicBundle` would be created that implements the pre-existing `Bundle` interface, but adds two functions.
+1. `BeforePlugins` to retrieve a list of plugins that should be run during before any other plugins
+2. `AfterPlugins` to retrieve a list of plugins that should be run after any other plugins
+
+Proposed implementation:
+
+File pkg/plugin/plugin.go:
+```go
+type DynamicBundle interface {
+	Bundle
+
+	// BeforePlugins returns the list of plugins that should be run before all plugins
+	BeforePlugins() []Plugin
+
+	// AfterPlugins returns the list of plugins that should be run after all plugins
+	AfterPlugins() []Plugin
+}
+```
+
+### Struct
+A new struct would be created that implements the `DynamicBundle` interface.
+
+Proposed implementation:
+
+File pkg/plugin/bundle.go:
+```go
+type dynamicBundle struct {
```

Comment:
What is the motivation for the name dynamic?
When/where/why this solution is dynamic?

Reply:
The idea behind the name dynamic is that there can be defaults that are defined to run before and after all plugins and the user can "inject" plugins into the chain. The idea is that there is one bundle definition for all plugins with a specific supported project  version rather than having to repeat the definition for each new language plugin.